src/util/container.h

```diff
@@ -305,3 +317,199 @@ class LRUCache
 	// we can't use std::deque here, because its iterators get invalidated
 	std::list<K> m_queue;
 };
+
+/*
+	Map that can be safely modified (insertion, deletion) during iteration
+	Caveats:
+	- you cannot insert null elements
+	- you have to check for null elements during iteration, those are ones already deleted
+	- size() and empty() don't work during iteration
+	- not thread-safe in any way
+*/
+
+template<typename K, typename V>
+class ModifySafeMap
+{
+public:
+	// this allows bare pointers but also e.g. std::unique_ptr
+	static_assert(std::is_default_constructible<V>::value,
+		"Value type must be default constructible");
+	static_assert(std::is_constructible<bool, V>::value,
+		"Value type must be convertible to bool");
+	static_assert(std::is_move_assignable<V>::value,
+		"Value type must be move-assignable");
+
+	typedef K key_type;
+	typedef V mapped_type;
+
+	ModifySafeMap() {
+		// the null value must convert to false and all others to true, but
+		// we can't statically check the latter.
+		assert(!null_value);
+	}
+	~ModifySafeMap() {
+		assert(!m_iterating);
+	}
+
+	// possible to implement but we don't need it
+	DISABLE_CLASS_COPY(ModifySafeMap)
+	ALLOW_CLASS_MOVE(ModifySafeMap)
+
+	const V &get(const K &key) const {
+		if (m_iterating) {
+			auto it = m_new.find(key);
+			if (it != m_new.end())
+				return it->second;
+		}
+		auto it = m_values.find(key);
+		return it == m_values.end() ? null_value : it->second;
+	}
+
+	void put(const K &key, const V &value) {
+		if (!value) {
+			assert(false);
+			return;
+		}
+		if (m_iterating)
+			m_new.emplace(key, value);
+		else
+			m_values.emplace(key, value);
+	}
+
+	void put(const K &key, V &&value) {
+		if (!value) {
+			assert(false);
+			return;
+		}
+		if (m_iterating)
+			m_new.emplace(key, std::move(value));
+		else
+			m_values.emplace(key, std::move(value));
+	}
+
+	V take(const K &key) {
+		V ret = V();
+		if (m_iterating) {
+			auto it = m_new.find(key);
+			if (it != m_new.end()) {
+				ret = std::move(it->second);
+				m_new.erase(it);
+			}
+		}
+		auto it = m_values.find(key);
+		if (it == m_values.end())
+			return ret;
+		ret = std::move(it->second);
+		if (m_iterating) {
+			it->second = V();
+			m_garbage++;
+		} else {
+			m_values.erase(it);
+		}
+		return ret;
+	}
+
+	bool remove(const K &key) {
+		return !!take(key);
+	}
+
+	size_t size() const {
+		if (m_iterating) {
+			// This is by no means impossible to determine, it's just annoying
+			// to code and we happen to not need this.
+			return unknown;
+		}
+		assert(m_new.empty());
+		if (m_garbage == 0)
+			return m_values.size();
+		size_t n = 0;
+		for (auto &it : m_values)
+			n += !it.second ? 0 : 1;
+		return n;
+	}
+
+	bool empty() const {
+		if (m_iterating)
+			return false; // maybe
+		if (m_garbage == 0)
+			return m_values.empty();
+		for (auto &it : m_values) {
+			if (!!it.second)
+				return false;
+		}
+		return true;
+	}
```

Comment:
Hmm hopefully not.
I've changed `::step` to avoid it at least.

One conceptual problem ist that you pay the cost of uncollected garbage every iteration and call to `size`/`empty` but garbage is only collected once you hit a certain limit...